[6.x] Fix loading deferred providers for binding interfaces and implementations

lprzybylek opened the pull request:
This PR is the effect of discussion in https://github.com/laravel/framework/issues/31588

It brings back behaviour from Laravel 5.7 when you could use deferred provides to bind an interface to implementation even when implementation had its own deferred provider. Changing `resolve` method to `make` results in calling `Illuminate\Foundation\Application::make` method which loads deferred providers. Calling `resolve` method calls `Illuminate\Container\Container::resolve` which does not load deferred providers. 

I'm targeting 6.x as it is stated in contribution guide, but in my opinion it may also be good to apply it to 5.8 for the sake of compatibility when migrating gradually from 5.7 to 5.8 and then to 6.x.

lprzybylek committed:
Add test that failes to load deffered provider of implementation while accessing class through interface

lprzybylek committed:
Call make instead of resolve to load deffered providers

This commit brings back behaviour from Laravel 5.7 when you could use deferred provides to bind an interface to implementation even when implementation had its own deferred provider. Changing `resolve` method to `make` results in calling `Illuminate\Foundation\Application::make` method which loads deferred providers. Calling `resolve` method calls `Illuminate\Container\Container::resolve` which does not load deferred providers.

driesvints commented on src/Illuminate/Container/Container.php:
```diff
@@ -261,7 +261,7 @@ protected function getClosure($abstract, $concrete)
             }
 
-            return $container->resolve(
+            return $container->make(
```
The make method is just an alias of resolve. This can't solve anything really.

lprzybylek replied:
It solves the issue. Try checking out previous commit where test was added. Resolve is an alias of make in `Illuminate\Container\Container`, but not in `Illuminate\Foundation\Application`.

taylorotwell replied:
@driesvints it actually could in the context of an application because the Foundation Application overrides `make` to do the deferred provider stuff.

driesvints commented on src/Illuminate/Container/Container.php:
```diff
@@ -620,13 +620,14 @@ public function makeWith($abstract, array $parameters = [])
      * @throws \Illuminate\Contracts\Container\BindingResolutionException
      */
-    public function make($abstract, array $parameters = [])
+    public function make($abstract, array $parameters = [], $raiseEvents = true)
```
Changing the method signature here is unfortunately a breaking change.

driesvints commented on tests/Foundation/FoundationApplicationTest.php:
```diff
@@ -473,6 +484,44 @@ public function register()
+{
+    public function register()
+    {
+        $this->app->bind(SampleInterface::class, SampleImplementation::class);
```
This simple can't work. You need to bind the primitive against the container otherwise it can't be resolved. I have no idea how this could ever have worked.

taylorotwell commented:
Why all the `raiseEvents` stuff? Is that necessary?

driesvints commented:
The problem here really is that the things happening in `Application::make` should probably also happen in `Application::resolve`. From a container perspective there shouldn't be a difference between `Container::make` and `Container::resolve`.

lprzybylek commented:
> Why all the `raiseEvents` stuff? Is that necessary?

To be honest I don't know how `$raiseEvents` affects the application. `getClosure` was using $raiseEvents = false and I didn't see any other way to run it with the same value than changing a signature and passing it over. Do you have any other ideas?

taylorotwell commented:
One solution may be to make the extra stuff in Application `make` to a new Application `resolve` method which calls that same code and then `parent::resolve`. That would probably let you do this without any breaking changes?

lprzybylek committed:
Revert "Call make instead of resolve to load deffered providers"

This reverts commit 8943aad90ba81c8a088b7e7c6cdd63ee41f6e01b.

lprzybylek committed:
Create resolve method which loads deferred providers if needed

It allows resolving object by interface when using separate deferred providers for interface and implementation.

lprzybylek committed:
Fix styling for phpdoc

lprzybylek committed:
Fix phpdoc

lprzybylek commented:
Good idea, I've updated the PR. Tests are OK and signatures are intact.
